crates/pathfinder/src/storage/state.rs

```diff
@@ -844,20 +763,32 @@ pub struct PageOfEvents {
 
 pub struct StarknetEventsTable {}
 impl StarknetEventsTable {
-    pub fn event_data_to_bytes(data: &[EventData]) -> Vec<u8> {
-        data.iter()
-            .flat_map(|e| (*e.0.as_be_bytes()).into_iter())
-            .collect()
+    pub fn encode_event_data_to_bytes(data: &[EventData], buffer: &mut Vec<u8>) {
+        buffer.extend(data.iter().flat_map(|e| (*e.0.as_be_bytes()).into_iter()))
     }
 
-    fn event_key_to_base64_string(key: &EventKey) -> String {
-        base64::encode(key.0.as_be_bytes())
+    fn encode_event_key_to_base64(key: &EventKey, buf: &mut String) {
+        base64::encode_config_buf(key.0.as_be_bytes(), base64::STANDARD, buf);
     }
 
-    pub fn event_keys_to_base64_strings(keys: &[EventKey]) -> String {
-        // TODO: we really should be using Iterator::intersperse() here once it's stabilized.
-        let keys: Vec<String> = keys.iter().map(Self::event_key_to_base64_string).collect();
-        keys.join(" ")
+    pub fn event_keys_to_base64_strings(keys: &[EventKey], out: &mut String) {
+        // with padding it seems 44 bytes are needed for each
+        let needed = (keys.len() * (" ".len() + 44)).saturating_sub(" ".len());
+        if let Some(more) = needed.checked_sub(out.capacity() - out.len()) {
+            out.reserve(more);
+        }
+
+        let _capacity = out.capacity();
+
+        keys.iter().enumerate().for_each(|(i, x)| {
+            Self::encode_event_key_to_base64(x, out);
+
+            if i != keys.len() - 1 {
+                out.push(' ');
+            }
+        });
+
+        debug_assert_eq!(_capacity, out.capacity());
```

Comment:
All of the reservations should be tight, so this will fire for example when you push two spaces instead of one, in tests.

Reply:
Your other comment just doesn't have a comment field for me underneath it.

Note that `reserve` is done first, so more space is asked if needed, but there's no correlation to the `needed` from `_capacity` nor the `out.capacity()` in the end. These asserts just check that the estimate was correct in the sense that it was enough. `std` can be rounding up to nearest power of two for example, and this will still be valid.

Reply:
Sure -- I think it was your comment that was misleading. Checking that the capacity is unchanged wouldn't detect pushing an extra space for example. But of course I don't think that's an issue.

Reply:
Oki I think we do need to discuss about this, I'll create the follow-up PR though first.